winch/codegen/src/isa/x64/masm.rs

```diff
@@ -193,6 +195,82 @@ impl Masm for MacroAssembler {
         self.asm.mul(src, dst, size);
     }
 
+    fn and(&mut self, dst: RegImm, lhs: RegImm, rhs: RegImm, size: OperandSize) {
+        let (src, dst): (Operand, Operand) = if dst == lhs {
+            (rhs.into(), dst.into())
+        } else {
+            panic!(
+                "the destination and first source argument must be the same, dst={:?}, lhs={:?}",
+                dst, lhs
+            );
+        };
+
+        self.asm.and(src, dst, size);
+    }
+
+    fn or(&mut self, dst: RegImm, lhs: RegImm, rhs: RegImm, size: OperandSize) {
+        let (src, dst): (Operand, Operand) = if dst == lhs {
+            (rhs.into(), dst.into())
+        } else {
+            panic!(
+                "the destination and first source argument must be the same, dst={:?}, lhs={:?}",
+                dst, lhs
+            );
+        };
+
+        self.asm.or(src, dst, size);
+    }
+
+    fn xor(&mut self, dst: RegImm, lhs: RegImm, rhs: RegImm, size: OperandSize) {
+        let (src, dst): (Operand, Operand) = if dst == lhs {
+            (rhs.into(), dst.into())
+        } else {
+            panic!(
+                "the destination and first source argument must be the same, dst={:?}, lhs={:?}",
+                dst, lhs
+            );
+        };
+
+        self.asm.xor(src, dst, size);
+    }
+
+    fn shift(&mut self, context: &mut CodeGenContext, kind: ShiftKind, size: OperandSize) {
+        let top = context.stack.peek().expect("value at stack top");
+
+        if size == OperandSize::S32 && top.is_i32_const() {
+            let val = context
+                .stack
+                .pop_i32_const()
+                .expect("i32 const value at stack top");
+            let reg = context.pop_to_reg(self, None, size);
+
+            self.asm.shift_ir(val as u8, reg, kind, size);
+
+            context.stack.push(Val::reg(reg));
+        } else if size == OperandSize::S64 && top.is_i64_const() {
+            let val = context
+                .stack
+                .pop_i64_const()
+                .expect("i64 const value at stack top");
+            let reg = context.pop_to_reg(self, None, size);
+
+            self.asm.shift_ir(val as u8, reg, kind, size);
+
+            context.stack.push(Val::reg(reg));
+        } else {
+            let rcx = context.gpr(regs::rcx(), self);
+            context.regalloc.free_gpr(rcx);
```

Comment:
Oops sorry, I mean to "start a review" not post a single comment.

Reply:
Can you clarify a bit?

Do you mean:
```rust
let rcx = context.gpr(regs::rcx(), self);
let src = context.pop_to_reg(self, Some(rcx), size);
let dst = context.pop_to_reg(self, None, size);
self.asm.shift_rr(src.into(), dst.into(), kind, size);
context.regalloc.free_gpr(src);
context.stack.push(Val::reg(dst));
```

For the else branch? That is, what's there now but not call `context.regalloc.free_gpr(rcx);`? Without that first line, I'm not clear on how I'm supposed to get `rcx`.

Reply:
(For some reason I don't get notifications for your replies, sorry for being late here)

I mean these lines particularly in the else branch:

```rust
let rcx = context.gpr(regs::rcx(), self);
context.regalloc.free_gpr(rcx);
// Number of bits to shift must be in the CL register.
let src = context.pop_to_reg(self, Some(rcx), size);
```

I think there's no need for the first two lines, since they are immediately allocating and freeing `rcx`; `context.pop_to_reg(self, Some(regs::rcx(), size)` will ensure that you get `rcx` allocated and assigned to whatever value is at the top of the stack as the shift factor. In the case of division it's a bit different because we need to allocate two registers upfront (quotient and remainder) and the reason we mark them as non-allocatable is because we want to make sure that the next call to `pop_to_reg` won't allocate any of those for the divisor (this is because the divisor comes first, but we don't want it to be allocated in `rdx:rax`).